### PR TITLE
add a method to collect the DNS names from a certificate

### LINKED ISSUE #64: Add method to get all the names from a certificate
Hi,
To integrate [rustls in sozu](https://github.com/sozu-proxy/sozu/issues/145), I'd need a way to get the common name and subject alternative names from a certificate, and apparently I should implement it on `EndEntityCert`. I opened a [related issue in rustls](https://github.com/ctz/rustls/issues/140).
If you're ok with this feature, I could work on a PR tomorrow.

### PR DESCRIPTION
**Pull request by Geal**
Fix #64

I am not too happy about it yet, I'd like to remove the `ToString` implementation for `DNSName`, since there's a `Into<&str>` for `DNSNameRef`.
Unfortunately, I have not been able to make the function return a vector of `DNSNameRef`yet.
For reference, the code I'm currently trying to build is:

```rust
#[cfg(feature = "std")]
pub fn list_cert_dns_names<'a>(cert: &'e super::EndEntityCert<'a>)
                                   -> Result<Vec<DNSNameRef<'a>>, Error> {
    let cert = &cert.inner;
    let names = std::cell::RefCell::new(Vec::new());

    iterate_names(cert.subject, cert.subject_alt_name, Ok(()), &|name| {
        match name {
            GeneralName::DNSName(presented_id) => {
                names.borrow_mut().push(DNSNameRef(presented_id));
            },
            _ => ()
        }
        NameIteration::KeepGoing
    }).map(|_| names.into_inner())
}
```

But there's a conflict between the lifetime of `name` (created in `iterate_names` and passed to the closure) and the lifetime of the `DNSNameRef` returned as result.

### CODE REVIEW

File: src/webpki.rs [hunk 1]
```diff
@@ -246,6 +246,19 @@ impl <'a> EndEntityCert<'a> {
...
+    /// Returns a list of the DNS names provided in the subject alternative names extension
+    ///
+    /// This function must not be used to implement custom DNS name verification.
+    /// Verification functions are already provided as `verify_is_valid_for_dns_name`
+    /// and `verify_is_valid_for_at_least_one_dns_name`.
+    ///
+    /// Requires the `std` default feature; i.e. this isn't available in
+    /// `#![no_std]` configurations.
+    #[cfg(feature = "std")]
+    pub fn list_dns_names(&self) -> Result<std::vec::Vec<DNSName>, Error> {
```
Comment (briansmith) [inline src/webpki.rs]: I think this should be called `dns_names()` instead of `list_dns_names()`.

File: tests/integration.rs [hunk 2]
```diff
@@ -107,3 +107,40 @@ fn time_constructor() {
...
+    let _ = expected_names.insert(String::from("release-stage.netflix.com"));
+    let _ = expected_names.insert(String::from("www.netflix.com"));
+
+    let mut dns_names = cert.list_dns_names()
+      .expect("should get all DNS names correctly for netflix end entity cert");
+
+    let names: std::collections::HashSet<String> = dns_names.drain(..).map(|name| {
+        let n: webpki::DNSName = name.into();
+        n.to_string()
+      }).collect();
```
Comment (briansmith) [inline tests/integration.rs]:
```rust
// Ensure that converting the list to a set doesn't throw away
// any duplicates that aren't supposed to be there
assert_eq!(actual_names.len(), expected_names.len());

let actual_names = HashSet::from_iter(actual_names.iter()
    .map(|dns_name| dns_name.as_ref().into())) // Maybe `into::<&str>` for clarity.
```

**Comment by briansmith** [comment]
> Unfortunately, I have not been able to make the function return a vector of DNSNameRef yet.

I will look into it now.

File: src/name.rs [hunk 3]
```diff
@@ -61,6 +64,13 @@ impl<'a> From<DNSNameRef<'a>> for DNSName {
     }
 }
 
+#[cfg(feature = "std")]
+impl std::string::ToString for DNSName {
```
[the remark this answers is not in the document]
Comment (Geal) [reply src/name.rs]: I removed it

File: src/name.rs [hunk 4]
```diff
@@ -136,6 +146,24 @@ pub fn verify_cert_dns_name(cert: &super::EndEntityCert,
     })
 }
 
+
+#[cfg(feature = "std")]
+pub fn list_cert_dns_names<'names>(cert: &super::EndEntityCert<'names>)
+                                   -> Result<Vec<DNSName>, Error> {
+    let cert = &cert.inner;
+    let names = std::cell::RefCell::new(Vec::new());
```
[the remark this answers is not in the document]
Comment (Geal) [reply src/name.rs]: I thought so too, but I get the following error when I do it: `error[E0387]: cannot borrow data mutably in a captured outer variable in an `Fn` closure`

File: tests/integration.rs [hunk 5]
```diff
@@ -107,3 +107,40 @@ fn time_constructor() {
...
+#[test]
+pub fn list_netflix_names()
+{
+    let ee = include_bytes!("netflix/ee.der");
+
+    let ee_input = untrusted::Input::from(ee);
+    let cert = webpki::EndEntityCert::from(ee_input)
+      .expect("should parse netflix end entity certificate correctly");
+
+    let mut expected_names = std::collections::HashSet::new();
```
[the remark this answers is not in the document]
Comment (Geal) [reply tests/integration.rs]: done

File: src/webpki.rs (hunk 1 above)
Comment (Geal) [reply src/webpki.rs]: done

File: tests/integration.rs (hunk 2 above)
Comment (Geal) [reply tests/integration.rs]: added

**Comment by Geal** [comment]
I'll add more integration tests as you suggested, then it should be good to go :)

**Comment by Geal** [comment]
the new tests are there, and the build failures are apparently unrelated to this PR (\*ring\* compilation failures on appveyor, and the beta toolchain not installed for the 2 xcode builds in travis).

**Comment by Geal** [comment]
hello, friendly ping on this :)
let me know if there's anything to change

**Comment by briansmith** [comment]
@Geal I don't know why I didn't comment on this here, but please see #66. At a minimum, we should have a test that demonstrates what happens when there is only a wildcard `*.example.com` and another test thta demonstrates what happens when there are both wildcard and non-wildcard subjectAltNames.

**Comment by briansmith** [comment]
The ideal state of wildcard support is hinted at in #66: Distinguish wildcard and non-wildcard entries with distinct enum variants of a new type. I would be willing to defer that ideal wildcard support to another PR but we at least need to decide to do when this function encounters a wildcard entry and we need the tests. It's probably less work to solve it all at once properly. WDYT?

**Comment by Geal** [comment]
What should be expected when there are both wildcard and non wildcard? Can
the wildcard match the non wildcard ? For sozu I will use the wildcard
entries as well.
I can go over the code next week

**Comment by briansmith** [comment]
Thanks for contributing this! This seems to have been subsumed by #91 so I'm optimistically closing this in favor of that one.

**Pull request closed without merging**